src/crimson/osd/heartbeat.cc

```diff
@@ -99,29 +99,19 @@ void Heartbeat::set_require_authorizer(bool require_authorizer)
 
 void Heartbeat::add_peer(osd_id_t peer, epoch_t epoch)
 {
-  auto [peer_info, added] = peers.try_emplace(peer);
-  auto& info = peer_info->second;
-  info.epoch = epoch;
-  if (added) {
-    logger().info("add_peer({})", peer);
-    auto osdmap = service.get_osdmap_service().get_map();
-    // TODO: use addrs
-    peer_info->second.con_front = front_msgr->connect(
-        osdmap->get_hb_front_addrs(peer).front(), CEPH_ENTITY_TYPE_OSD);
-    peer_info->second.con_back = back_msgr->connect(
-        osdmap->get_hb_back_addrs(peer).front(), CEPH_ENTITY_TYPE_OSD);
-  }
+  auto [iter, added] = peers.try_emplace(peer, *this, peer);
+  auto& peer_info = iter->second;
+  peer_info.set_epoch(epoch);
```

Comment:
Understood, the original logic is simple enough because PeerInfo is removed/added as a whole during reset event.
I think it becomes necessary if there's a need to extend PeerInfo to resolve deadlocks/racing:
- help me understand the possible external operations/events for PeerInfo and how its state changes internally;
- help reviewers to better understand the introduced changes with new external events and internal session managements;

Reply:
what is the deadlocks/racing you are trying to resolve?  what complicates the originally simple logic? i still don't get why we need to extract PeerInfo into an individual class.

as always, it'd be ideal if you could add the reason of the change in commit message.

Reply:
> what is the deadlocks/racing you are trying to resolve? ... as always, it'd be ideal if you could add the reason of the change in commit message.

Please refer to the original analysis in https://github.com/ceph/ceph/pull/33909#issuecomment-599380821 for the racing/deadlock/compatibility issues, and the solution proposed by commit https://github.com/ceph/ceph/pull/34311/commits/382b3718e2f5007dd9c55b17b1f0334f540bc6ff, where there's an explanation of the solutions in its commit message. I can extend that message according to our discussions.

> ... what complicates the originally simple logic?  i still don't get why we need to extract PeerInfo into an individual class. ...

The major reason is that **the single connection feature** introduces 3 issues to the originally simple logic adopted from the classic OSD's 2-connection solution.

Basically, in order to solve the issue 1 (racing), the `Peer`(renamed from `PeerInfo`) attempts to identify if the racing is happening, this is because the heartbeat peer is not necessarily to be symmetric. If racing is actually happening, `Peer` then decides to wait for another side if it loses, or connect proactively if it wins (`whoami > peer`).

To solve the issue 2 (deadlock), the direct impact is that for a reset event from either `con_front` or `con_back`, the heartbeat class should not simply remove the related `PeerInfo` which contains both of them. Instead, `Peer` is improved(complicated) to be a session-alike class which remembers/tracks individual front/back connectivity. And it only starts to track from a clean `ping_history` (same with the original logic) when both are connected, and stop tracking as soon as either of them is reset.

The session-alike `Peer` class is also needed to solve issue 1, because compared with the "originally simple logic", the connectivity status needs to be tracked explicitly in order to control the racing related decisions before connected.

To deal with issue 3 (compatibility), I think the most important thing is to keep healthy/unhealthy check identical to the original logic, because it is working for a long time. This is the reason to introduce a related clean-up (https://github.com/ceph/ceph/pull/34311/commits/32ff43fbfac739b934f26f3037fee40054fe7068) and keep this logic mostly unchanged in commit https://github.com/ceph/ceph/pull/34311/commits/382b3718e2f5007dd9c55b17b1f0334f540bc6ff.

Reply:
> but i don't think its existence can prove its correctness or efficiency.

Agree, but the difficult part is that the "correctness" is not very obvious to me without fully understand the original health check logic, especially when there's a twilight state between healthy and unhealthy.

I think we indeed have room for further improvements to health check logic, but it would be more understandable after the clean-up.